Settings page added

Nppatel97 opened the pull request:
New Settings page made. @AhmedNSidd if you can add a link or something for this so I can view this page while modifying, that would be dope! Also, there are only two settings that I am aware of right? To honor or not honor availability and ability to reschedule. Also, this is literally just a copy pasta code of agenda.html so I haven't changed anything except the title which says 'Settings | Schedulearn'.

Nppatel97 commented:
> Looks good to me, but I wonder if there is any way to share code/reuse components from existing files instead of copy 🍝 code... any thoughts?

That's a great idea! We can make a skeleton file and reuse it in all other files but it would be kinda difficult to do in HTML.. Something like React would be awesome for that. Or if there is some feature in Python or JS that I may not be aware of

AhmedNSidd commented:
I am actually working on a branch right now that's doing something similar to what you're asking. Specifically, it's setting up css files that are generic to all pages to minimize reuse of code. It's taking some time to work on it because there's a lot of variability across the css files. Anyway @Nppatel97, I pushed some code up that has set up the routing for the settings page. So you can see the page now at http://localhost:8000/dashboard/settings/ 
If you have any questions lmk

Nppatel97 commented:
Alright guys, Settings page is complete with all the links and options in place. My Profile link is moved to primary navbar and is replaced by settings. Will have to do this in all other files though. Now just need to add the back-end logic to this. Here are some images. Let me know what you think and any changes that it might need @AhmedNSidd @sdevalapurkar @DSchriemer :)

![settings-1](https://user-images.githubusercontent.com/22160794/59754390-da50d600-923a-11e9-9796-eadd3b3e7ce8.png)
![settings-2](https://user-images.githubusercontent.com/22160794/59754392-dae96c80-923a-11e9-9923-68ca8966197f.png)

AhmedNSidd commented:
That looks absolutely fantastic man. Exactly how I wanted it to look. GREAT job. Although for the setting titles, I think we should capitalize the first letter of each word. After work, I'm gonna get home and check out how this looks in the browser and start implementing some of the backend into this

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,11 @@
+$(document).ready(function () {
```
can we use ES6 style arrow functions here?
```
$(document).ready(() => {

});
```

Nppatel97 replied:
For sure!

sdevalapurkar commented on schedule_lessons/dashboard/urls.py:
```diff
@@ -38,5 +38,6 @@
     path('unblock/<user_id>', views.unblock_user, name='unblock_user'),
-    path('save_tutorial_preferences', views.save_tutorial_preferences, name='save_tutorial_preferences')
+    path('save_tutorial_preferences', views.save_tutorial_preferences, name='save_tutorial_preferences'),
+    path('settings/', views.settings, name='settings')
```
why `settings/` instead of just `settings`?

AhmedNSidd replied:
Good point. Actually, the urls above settings are inconsistent, they should have trailing slashes as well. I opened up issue #138, so we can fix this in another PR.

sdevalapurkar commented on schedule_lessons/dashboard/views.py:
```diff
@@ -895,6 +895,21 @@ def save_tutorial_preferences(request):
+    notifications = Notification.objects.filter(
+        user=request.user).order_by("-created_on")
+    context = {
+        "notifications": []
```
question: how come context doesn't have a property called `unread_notifications` even tho it is being used later on?

AhmedNSidd replied:
True actually, I opened up issue #139 about this so we can clean this up later since there are other instances where I do this in this file.

DSchriemer commented:
Looking very good, nice work!!

Nppatel97 commented:
> I had a chance to look at this after work. I have a few comments on things that can be improved a bit. Firstly I think the Settings Title in the webpage should be bigger (maybe make it h3 or even bigger, use your best judgement). Make the setting names and descriptions relatively bigger too (idk if you made them small so it looks better on mobile, if that's the case, let me know).
> 
> I think for the future of the settings page, I wanna move the "Blocked Users", "Change Password", and "Delete Account" options from the profile page to the settings page. It's not something I think we should do in this PR but maybe we should setup issues to tackle these issues in future versions (or if we get the chance, v1.0).

Yeah I wanted to make them look consistent with the text in navbars so I chose these sizes for Settings title and options. I'll fiddle around with sizes and see which ones look better on both mobile and desktop.

And yes, I think those options would be better in settings (than in my profile page), I think it could be easy to do it in v1.0 as long as there is not much back-end logic involved. (Maybe even in this PR?)

Nppatel97 commented:
> Looking very good, nice work!!

Thanks! Let me know if you have any ideas on improving this UI/UX :)

Nppatel97 commented:
I have updated all the navbar links so now they all show my profile in primary navbar and settings under profile pic. Also moved the change password, blocked users and delete account options to the settings page. Not sure how it connects to the back-end so I am leaving that for @AhmedNSidd. 

![image](https://user-images.githubusercontent.com/22160794/60381752-9914af00-9a0e-11e9-9184-fe7c6102e806.png)

Also, the secondary navbar of public_profile.html has not been optimized by me in past because there are a lot (a lotttt) of if-else django stuff in it as shown in the image ↓

![image](https://user-images.githubusercontent.com/22160794/60381742-7bdfe080-9a0e-11e9-93ac-88e0e35fc506.png)

Check the lastest commit and let me know of any bugs or if I missed anything. Cheers! 🍻

AhmedNSidd commented:
@Nppatel97 Please check out this latest commit and make sure it looks alright. I used the chrome coverage tool to get rid of any unused css/js code for the settings page.

AhmedNSidd commented on schedule_lessons/dashboard/static/dashboard/css/agenda.css:
```diff
@@ -395,48 +416,44 @@ html body {
 
 #picturePrompt {
-  height: 300px;
+  width: 90%;
```
I am guessing you made this change to look better on mobile, but the image was too huge on laptop. Please check out my commit on "Fixing the huge picture prompt" and let me know if the changes I made look alright

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+
+function getCookie(name) {
+  var cookieValue = null;
+  if (document.cookie && document.cookie != '') {
```
it is always a good idea to **fail fast** in programming
so let's use that approach here. is it intended that this function return `null` when there is no cookie value??

if yes, let's just do:

```
if (!document.cookie) {
    return null;
}

// here do rest of computation
```

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+    for (var i = 0; i < cookies.length; i++) {
+      var cookie = jQuery.trim(cookies[i]);
+      // Does this cookie string begin with the name we want?
+      if (cookie.substring(0, name.length + 1) == (name + '=')) {
```
always use `===` because it does type checking as well
can get messed up in js if using `==` without knowing it

sdevalapurkar replied:
example:
```
if (0 == false) {
    //this conditional is true in js even tho 0 is a number type and false is a boolean
}
```

```
if (0 === false) {
    //this is not true because of type checks which is safer and what we want
}
```

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+  if (document.cookie && document.cookie != '') {
+    var cookies = document.cookie.split(';');
+    for (var i = 0; i < cookies.length; i++) {
+      var cookie = jQuery.trim(cookies[i]);
```
no vars

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+}
+
+$.ajaxSetup({
+  beforeSend: function (xhr, settings) {
```
arrow function

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+
+$.ajaxSetup({
+  beforeSend: function (xhr, settings) {
+    if (!(/^http:.*/.test(settings.url) || /^https:.*/.test(settings.url))) {
```
not sure what is being checked in this if statement, could you please clarify, thanks!

AhmedNSidd replied:
I changed the if condition, let me know if the new one makes sense to you.

AhmedNSidd commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+    $.ajax({
+      type: 'POST',
+      url: '/dashboard/my_profile/change_password/',
+      data: { 'old_password': $("#oldPasswordInput").val(), 'new_password1': $("#newPasswordInput1").val(), 'new_password2': $("#newPasswordInput2").val() },
```
For the first point, from what I've read, keys need to be defined as strings. For the second point, that's a good point. For simple errors like empty forms, frontend checking makes more sense. I actually think we should have both frontend and backend checking. Frontend for the simple stuff and backend for more complicated backend checking. I started an issue on this #141. If you have good resources to share on error handling in webapps, pop them in there.

sdevalapurkar replied:
all the checks should happen in both backend and frontend
frontend for user experience improvements and also reducing wait times for server responses
backend for security and actual functionality purposes in case a user tries to make a mock request using curl or something like that to an endpoint without actually using the frontend and tries to change someone's password or delete someone's account

AhmedNSidd commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+        $('.error-list').html('Something went wrong, please try again');
+      },
+      success: function (response) {
+        if (response.status_code == 200) {
```
Yeah it is possible to move have the error callback handle the error status codes. I was actually unaware as to how to do this with AJAX/Django which is why the code is messed up like this. Opened up issue #142 to fix this, and all other instances of things like this in the codebase.

sdevalapurkar replied:
okay cool, yeah that would be good

AhmedNSidd commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+
+  $('#notificationsDropdownLink').click(() => {
+    var attr = $('#notificationIcon').attr('data-count');
+    if (typeof attr !== typeof undefined && attr !== false) {
```
Yup you're right, nice spot.

AhmedNSidd commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -0,0 +1,88 @@
+    $.ajax({
+      url: '/dashboard/clear_notifications/',
+      type: 'post',
+      error: function (xhr, status) {
```
Removed them.

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -46,43 +45,47 @@ $(document).ready(() => {
+
+  // Search through the cookies to find the desired cookie.
+  let cookieValue = null;
+  let cookies = document.cookie.split(';');
```
pre sure `cookies` can be a `const` i dont see it being updated anywhere
```const cookies = document.cookie.split(';');```

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -55,6 +55,15 @@ $(document).ready(() => {
   });
+
+  $(".preference").click((event) => {
+    const preference_id = event.target.id;
```
since we are only using a few properties of this event object, let's try this:

```
const { id, checked } = event.target;
```

and then use `id` and `checked`

sdevalapurkar commented on schedule_lessons/dashboard/static/dashboard/js/settings.js:
```diff
@@ -55,6 +55,15 @@ $(document).ready(() => {
+    const preference_id = event.target.id;
+    $.ajax({
+      type: "POST",
+      url: "/dashboard/modify_preference/" + preference_id + "/",
```
url: ``/dashboard/modify_preference/${id}/``

sdevalapurkar replied:
use the string template literal syntax
